source/dhtslib/sam.d

```diff
@@ -728,6 +728,118 @@ private char* reverse(char* str)
     return str;
 }
 
+struct SamWriter{
+    /// filename; as usable from D
+    string filename;
+
+    /// filename \0-terminated C string; reference needed to avoid GC reaping result of toStringz when ctor goes out of scope
+    private immutable(char)* fn;
+
+    /// htsFile
+    private htsFile* fp;
+
+    /// hFILE if required
+    private hFILE* f;
+
+    /// header struct
+    bam_hdr_t* header = null;
+
+    private kstring_t line;
+
+    /// disallow copying
+    @disable this(this);
+
+    /** Create a representation of SAM/BAM/CRAM file from given filename or File
+
+    Params:
+        fn =            string filename (complete path passed to htslib; may support S3:// https:// etc.)
+        extra_threads = extra threads for compression/decompression
+                        -1 use all available cores (default)
+                        0  use no extra threads
+                        >1 add indicated number of threads (to a default of 1)
+    */
+    this(T)(T fn,bam_hdr_t * header, int extra_threads = -1)
+    if (is(T == string) || is(T == File))
+    {
+        import std.parallelism : totalCPUs;
+
+        // open file
+        static if (is(T == string))
+        {
+            this.filename = fn;
+            this.fn = toStringz(fn);
+            this.fp = hts_open(this.fn, cast(immutable(char)*) "w");
+        }
+        else static if (is(T == File))
+        {
+            this.filename = fn.name();
+            this.fn = toStringz(fn.name);
+            this.f = hdopen(fn.fileno, cast(immutable(char)*) "w");
+            this.fp = hts_hopen(this.f, this.fn, cast(immutable(char)*) "r");
+        }
+        else assert(0);
+
+        if (extra_threads == -1)
```

Comment:
If this code is identical in SAMReader , should be factored out to ease maintenance

Reply:
We could make it a mixin? But that might be less tenable. Or make it a function that operates on the hts file pointer.

Reply:
Definitely not mixin. 
Another approach is to make SAMReader and SAMWriter classes inherit frmo SAMFile, since the constructors are literally identical. What will we break if we make them classes instead of structs?

Reply:
literally identical except the file open mode flag i mean

Reply:
If I change them to classes it doesn't appear to break anything severely when unit-testing other than a few fixes to include ```new``` and remove ```&(this)```. We will have to change all instantiations to include the ```new``` keyword. 
I can push changes to demonstrate. If we decide we don't want to do that we can revert.

Reply:
it makes the most sense from a code deduplication perspective.

for tings thatwe will allocate very few of, like readers/writers, it doesnt really matter if its on stack or heap

Reply:
Open issue for code dedupliation and i'll let thru as is